### ISSUE TITLE
Opening notebooks with alternative content managers

### ISSUE THREAD

**Issue by grst** [jupyterlab/jupyterlab#3896]
There are several projects that provide alternative `ContentManagers` which support reading and storing JupyterNotebooks to alternative formats (e.g. markdown), for instance

* [ipymd](https://github.com/rossant/ipymd)
* [notedown](https://github.com/aaren/notedown)

With one of these plugins installed, Jupyter notebook would open a markdown file as a notebook. Jupyter Lab keeps opening markdown files in the text editor, although the `ContentManager` seems to work in principle:

* the icon of markdown files changes to 'Notebook'
* when creating a new notebook and renaming it to have a `.md` extension, it is correctly saved as markdown file. 

**What would be the best way, to make these projects compatible with jupyter lab?
Is it hardcoded somewhere that only `ipynb` files are opened as notebook?** 

See also https://github.com/rossant/ipymd/issues/88

**Comment by timkpaine** [comment]
you can just overload the default contents manager in your config json

**Comment by timkpaine** [comment]
Now if you want them to live *alongside* the default contents manager, thats another question, and will require changes to those packages.

**Comment by timkpaine** [comment]
To live alongside, you should inherit from notebook.services.contents.manager.ContentsManager, but modify the `get` and `save` to handle ".md" as notebook. Look [here](https://github.com/jupyter/notebook/blob/master/notebook/services/contents/filemanager.py#L418) for more details.

**Comment by grst** [comment]
Sorry if I have not been clear. 
The problem is that despite adding the overloaded content manager, `.md` files still get opened as text and not as notebook.

**Comment by blink1073** [comment]
Ah, that would require a JupyterLab extension that added a [model factory](https://github.com/jupyterlab/jupyterlab/blob/5e1863aafb4c4535c8e26029316ac2415adcc185/packages/docregistry/src/registry.ts#L921) that translated `.md` content and create a notebook model.  It would also require [widget factory](https://github.com/jupyterlab/jupyterlab/blob/5e1863aafb4c4535c8e26029316ac2415adcc185/packages/docregistry/src/registry.ts#L879) that uses the new model model factory.

**Comment by mwouts** [comment]
Hello, I have the same question as @grst above. There are a few content manager classes around that allow to open Rmd or md notebooks in classical jupyter by simply installing one of those python packages, and modifying jupyter notebook config to
```
# notedown: https://github.com/aaren/notedown
c.NotebookApp.contents_manager_class = 'notedown.NotedownContentsManager'

# OR ipymd: https://github.com/grst/ipymd
c.NotebookApp.contents_manager_class = 'ipymd.IPymdContentsManager'
c.IPymdContentsManager.format = 'rmarkdown'

# OR nbrmd: https://github.com/mwouts/nbrmd
c.NotebookApp.contents_manager_class = 'nbrmd.RmdFileContentsManager'
c.ContentsManager.pre_save_hook = 'nbrmd.update_rmd_and_ipynb'
```

In the case of Jupyter lab we have seen that the alternative content manager is indeed the one used by jupyter lab. For instance, in the case of `nbrmd`, which offers to maintain up-to-date `.Rmd` and `.ipynb` versions of the notebook, the alternative version gets created or updated when the `.ipynb` is saved.

However, we still miss the possibility to open Rmd and md files as notebooks in jupyter lab (they open as text files).

@blink1073 , I've seen your comment about developping a jupyter lab extension. Unfortunately this is something I've never done before and I have no idea how to proceed. If I summarise the above, we just need the extension to tell jupyter lab that .md and .Rmd files are notebooks, right ? That kind of extension should be among the easiest, isn't it ? If you had an example of such an extension, that would be of much help.

**Comment by mwouts** [comment]
Well, I just noticed that when I use my own ContentsManager class, the Rmd notebooks have a notebook icon. Still when I click they open as text. Is this expected? Or is the information that Rmd are notebooks not propagated to the right place? Thanks

Empty `jupyter_notebook_config.py` | `c.NotebookApp.contents_manager_class = 'nbrmd.RmdFileContentsManager'`
:-----------------------:|:-----------------------:
![image](https://user-images.githubusercontent.com/29915202/42063464-1c144a20-7b32-11e8-94be-ea3bb5f98f5e.png) | ![image](https://user-images.githubusercontent.com/29915202/42063477-333f5bb8-7b32-11e8-884c-8e2ca62b9546.png)

**Comment by jasongrout** [comment]
FYI, here are some quick code links to relevant lines of code in JuptyerLab:

Here is where the notebook file type is defined: https://github.com/jupyterlab/jupyterlab/blob/30fd2bdadeee39170ac6ebf1a6766c9d9832c292/packages/docregistry/src/registry.ts#L1069

There are more filetypes defined below: https://github.com/jupyterlab/jupyterlab/blob/30fd2bdadeee39170ac6ebf1a6766c9d9832c292/packages/docregistry/src/registry.ts#L1097

Here is where we tell JupyterLab that the notebook object is used for 'notebook' file types, and that it is indeed the default opener for 'notebook' file types: https://github.com/jupyterlab/jupyterlab/blob/c4cdb84b91d41ea825a7eb1bcd92d1d2b5c161c5/packages/notebook-extension/src/index.ts#L412

**Comment by jasongrout** [comment]
I mean, it sounds like basically you want to add some more extensions to that `notebook` file type. I'm not sure we have a clean way to do that, but perhaps what you can do is in an extension query the filetypes that are available, find the notebook one, and add your preferred extensions to it.

**Comment by grst** [comment]
Closed in favor of #4924
